go/logic/applier.go

```diff
@@ -566,14 +566,41 @@ func (this *Applier) StartSlaveSQLThread() error {
 	return nil
 }
 
+func (this *Applier) isReplicationStopped() bool {
+	query := `show slave status`
+	replicationStopped := false
+
+	err := sqlutils.QueryRowsMap(this.db, query, func(rowMap sqlutils.RowMap) error {
+		replicationStopped = rowMap["Slave_IO_Running"].String == "No" && rowMap["Slave_SQL_Running"].String == "No"
+		return nil
+	})
+
+	if err != nil {
+		return false
+	}
+	return replicationStopped
+}
+
 // StopReplication is used by `--test-on-replica` and stops replication.
 func (this *Applier) StopReplication() error {
-	if err := this.StopSlaveIOThread(); err != nil {
-		return err
-	}
-	if err := this.StopSlaveSQLThread(); err != nil {
-		return err
+	if this.migrationContext.ManualReplicationControl {
+		for {
+			log.Info("Waiting for replication to stop...")
+			if this.isReplicationStopped() {
+				log.Info("Replication stopped.")
+				break
+			}
+			time.Sleep(5 * time.Second)
+		}
```

Comment:
We discussed the use of plugins. In such case the plugin would be called synchronously and would stop replication.
The way I see it there is no need to wait for replication to stop: I'm assuming your call to amazon to stop replication is also synchronous, i.e. when the plugin returns, replication is stopped (or there is an error).
In either case, waiting is not the desired behavior.

I would suggest the PR could be made much smaller: merely introducing the flag; and if it is set, we skip the `StopSlave...` operation.

Later on I will add the plugin injection.

WDYT?